crates/p2p/src/swarm/message.rs

```diff
@@ -0,0 +1,141 @@
+
+/// Protocol identifier for setup messages
+pub(crate) const SETUP_PROTOCOL_ID: u8 = 1;
+
+/// Wrapper for signed messages
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct SignedMessage {
+    /// The serialized message content
+    pub message: Vec<u8>,
+    /// The signature of the message
+    pub signature: Vec<u8>,
+}
+
+impl SignedMessage {
+    /// Creates a new signed message with the given signer
+    pub(crate) fn new<T, S>(
+        message: T,
+        signer: &S,
+        app_public_key: PublicKey,
+    ) -> Result<Self, Box<dyn std::error::Error + Send + Sync>>
+    where
+        T: Serialize,
+        S: crate::signer::ApplicationSigner,
+    {
+        let message_bytes = serde_json::to_vec(&message)
+            .map_err(|e| Box::new(e) as Box<dyn std::error::Error + Send + Sync>)?;
+        let signature = signer.sign(&message_bytes, app_public_key)?;
+
+        Ok(Self {
+            message: message_bytes,
+            signature,
+        })
+    }
+
+    /// Verifies the signature of this message
+    pub(crate) fn verify_signature(
+        &self,
+        app_public_key: &PublicKey,
+    ) -> Result<bool, Box<dyn std::error::Error + Send + Sync>> {
+        Ok(app_public_key.verify(&self.message, &self.signature))
+    }
+
+    /// Deserializes the inner message
+    pub(crate) fn deserialize_message<T>(&self) -> Result<T, serde_json::Error>
+    where
+        T: for<'de> serde::Deserialize<'de>,
+    {
+        serde_json::from_slice(&self.message)
+    }
+
+    /// Deserializes a signed message from JSON bytes
+    #[expect(dead_code)]
+    pub(crate) fn from_json_bytes(data: &[u8]) -> Result<Self, serde_json::Error> {
+        serde_json::from_slice(data)
+    }
+}
+
+impl SignedMessage {
+    /// Creates a new signed setup message with the given signer
+    pub(crate) fn new_signed_setup<S: crate::signer::ApplicationSigner>(
+        app_public_key: PublicKey,
+        local_peer_id: PeerId,
+        remote_peer_id: PeerId,
+        signer: &S,
+    ) -> Result<Self, Box<dyn std::error::Error + Send + Sync>> {
+        let setup_message =
+            SetupMessage::new(app_public_key.clone(), local_peer_id, remote_peer_id);
+
+        SignedMessage::new(setup_message, signer, app_public_key)
+    }
+}
+
+/// Setup message structure for the handshake protocol.
+/// Now serialized/deserialized using JSON instead of custom binary format.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub(crate) struct SetupMessage {
+    /// Protocol version
+    pub version: u8,
+    /// Protocol identifier (setup)
+    pub protocol: u8,
+    /// The application public key (Ed25519) - stored as bytes for serialization
+    pub app_public_key: Vec<u8>,
+    /// Local transport ID (PeerId) - stored as bytes for serialization
+    pub local_transport_id: Vec<u8>,
+    /// Remote transport ID (PeerId) - stored as bytes for serialization
+    pub remote_transport_id: Vec<u8>,
```

Comment:
Same as above. I really want to avoid `Vec<u8>`s and just use these raw bytes for the messages that the implementers will define when using `strata-p2p`.